entity-framework/core/what-is-new/ef-core-7.0/whatsnew.md

```diff
@@ -2064,3 +2064,558 @@ Configuration of stored procedures many-to-many join entities can be performed a
+
+EF Core already has built-in support for injecting some special services into context instances; for example, see [Lazy loading without proxies](xref:core/querying/related-data/lazy#lazy-loading-without-proxies), which works by injecting the `ILazyLoader` service.
+
+An `IMaterializationInterceptor` can be used to generalize this to any service. The following example shows how to inject an <xref:Microsoft.Extensions.Logging.ILogger> into entities such that they can perform their own logging.
+
+> [!NOTE]
+> Injecting services into entities couples those entity types to the injected services, which some people consider to be an anti-pattern.
+
+As before, an interface is used to define what can be done.
+
+<!--
+    public interface IHasLogger
+    {
+        ILogger? Logger { get; set; }
+    }
+-->
+[!code-csharp[IHasLogger](../../../../samples/core/Miscellaneous/NewInEFCore7/InjectLoggerSample.cs?name=IHasLogger)]
+
+And entity types that will log must implement this interface. For example:
+
+<!--
+    public class Customer : IHasLogger
+    {
+        private string? _phoneNumber;
+
+        public int Id { get; set; }
+        public string Name { get; set; } = null!;
+
+        public string? PhoneNumber
+        {
+            get => _phoneNumber;
+            set
+            {
+                Logger?.LogInformation(1, $"Updating phone number for '{Name}' from '{_phoneNumber}' to '{value}'.");
+
+                _phoneNumber = value;
+            }
+        }
+
+        [NotMapped]
+        public ILogger? Logger { get; set; }
+    }
+-->
+[!code-csharp[CustomerIHasLogger](../../../../samples/core/Miscellaneous/NewInEFCore7/InjectLoggerSample.cs?name=CustomerIHasLogger)]
+
+This time, the interceptor must implement `IMaterializationInterceptor.InitializedInstance`, which is called after every entity instance has been created and its property values have been initialized. The interceptor obtains an `ILogger` from the context and initializes `IHasLogger.Logger` with it:
+
+<!--
+    public class LoggerInjectionInterceptor : IMaterializationInterceptor
+    {
+        private ILogger? _logger;
+
+        public object InitializedInstance(MaterializationInterceptionData materializationData, object instance)
+        {
+            if (instance is IHasLogger hasLogger)
+            {
+                _logger ??= materializationData.Context.GetService<ILoggerFactory>().CreateLogger("CustomersLogger");
+                hasLogger.Logger = _logger;
+            }
+
+            return instance;
+        }
+    }
+-->
+[!code-csharp[LoggerInjectionInterceptor](../../../../samples/core/Miscellaneous/NewInEFCore7/InjectLoggerSample.cs?name=LoggerInjectionInterceptor)]
+
+This time a new instance of the interceptor is used for each `DbContext` instance, since the `ILogger` obtained can change per `DbContext` instance, and the `Ilogger` is cached on the interceptor:
```

Comment:
```suggestion
This time a new instance of the interceptor is used for each `DbContext` instance, since the `ILogger` obtained can change per `DbContext` instance, and the `ILogger` is cached on the interceptor:
```